### PR TITLE
Added Fathead for React

### LINKED ISSUE #473: JavaScript: Create New ReactJS Fathead
# ReactJS Fathead

Contribute to DuckDuckGo!

screen shot 2016-10-21 at 16 36 35
### What you will need to know?

Either **JavaScript**, Python, Ruby, or Perl for Fatheads. Bash is also useful. You will download the reference source and create an **output.txt** file containing the parsed documentation. Read more about Fatheads here: 
https://docs.duckduckhack.com/resources/fathead-overview.html
### The source :rocket:

https://facebook.github.io/react/docs/top-level-api.html
https://facebook.github.io/react/docs/component-api.html
### The goal

Help _[improve JavaScript language search results](https://forum.duckduckhack.com/t/javascript-search-overview)_ on DuckDuckGo as part of our Programming Mission! We're trying to reach 100% Instant Answer (IA) coverage for JavaScript-related searches by creating new Instant Answers or improving existing ones.

Code to some relevant examples:
- [jQuery Reference](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/jquery)
- [NodeJS Reference](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/node_js)
- [MDN JS Reference](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/mdnjs)

**Fathead Example (jQuery Reference)**

screen shot 2016-10-21 at 17 06 36
### The steps
- [x] Claim this issue (Comment Below)
- [x] Please review our [Contributing Guide](https://github.com/duckduckgo/zeroclickinfo-fathead/blob/master/CONTRIBUTING.md)
- [x] Set up codio environment (See: https://docs.duckduckhack.com/welcome/setup-dev-environment.html) and fork this repository
- [x] Create the Fathead
- [x] Create a PR
- [x] Ping @bfmags for a review
### Questions?
- [Documentation](https://docs.duckduckhack.com/)
- [The Forum](https://forum.duckduckhack.com/c/programming/javascript)
- Join [DuckDuckHack Slack](https://quackslack.herokuapp.com/) 
- Ping one of the [Community Leaders](https://duck.co/help/community/community-leaders) or @bfmags

---

Instant Answer Page: https://duck.co/ia/view/react_api_docs

_Thanks to @pjhampton for the template_

### PR DESCRIPTION
**Pull request by dip-kush**
## Type of Change

- [x] New Instant Answer
- [ ] Improvement
    - [ ] Bug fix
    - [ ] Enhancement
- [ ] Non–Instant Answer
    - [ ] Other (Role, Template, Test, Documentation, etc.)

## Related Issues and Discussions

Fixes #473

## People to notify

@pjhampton 
@moollaza 

---

IA Page: https://duck.co/ia/view/reactjs

Forum Topic: [Create IA: ReactJS Fathead](https://forum.duckduckhack.com/t/create-ia-reactjs-fathead/999)

### CODE REVIEW

**Comment by daxtheduck** [comment]
@dip-kush Thanks for the pull request! Unfortunately, we're only accepting contributions and improvements to a limited portion of the search space.
            
The staff and community are currently focused on making DuckDuckGo the [best search engine for programmers](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53). To ensure we're making as much progress as possible, **we are only accepting Instant Answers and updates related to the programming mission**.
    
If this is not related to the Programming Mission, it will likely be closed but you can:
    
1. Help us prevent further confusion by telling us where you found out about DuckDuckHack or where we could've alerted you earlier.
2. Keep this code handy! We'll be expanding into other search topics in the future!
3. Join us at our new home: https://forum.duckduckhack.com/

If this is related to the Programming Mission, it will be reviewed by staff and the community as soon as possible!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by bfmags** [comment]
@dip-kush Looking good! Nice work!

Just a small change and a code update :

- [ ] Change folder name to "react_js"

- [ ] updated to use the new fathead templates (example below)

```html
<section class="prog__container">
    <p>Creates a JavaScript Date instance that represents a single moment in time.</p>
    <pre>
        <code>
            new Date;
            new Date(value);
            new Date(dateString);
            new Date(year, month[, day[, hour[, minutes[, seconds[, milliseconds]]]]]);
        </code>
    </pre>
    <span class="prog__sub">Return Value</span>
    <p>The removed element.</p>
</section>
```

Thanks!

File: lib/fathead/react/parse.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+                example = self.parse_example(api)
+
+                section = {
+                    'title': title,
+                    'href': href,
+                    'content': content,
+                    'example': example
+                }
+                self.parsed_data.append(section)
```
Comment (bfmags) [inline lib/fathead/react/parse.py]: Looks good :) just needs update to use the new fathead templates (example below)

```html
<section class="prog__container">
    <p>Creates a JavaScript Date instance that represents a single moment in time.</p>
    <pre>
        <code>
            new Date;
            new Date(value);
            new Date(dateString);
            new Date(year, month[, day[, hour[, minutes[, seconds[, milliseconds]]]]]);
        </code>
    </pre>
    <span class="prog__sub">Return Value</span>
    <p>The removed element.</p>
</section>
```

**Comment by dip-kush** [comment]
Made Changes according to the review:
1. Changed folder name to react_js
2. Updated the code to use the new fathead templates

**Comment by daxtheduck** [comment]
@dip-kush this is now merged! It should be live on DuckDuckGo within one week (or during the next release) and we'll be sure to notify you when it happens.

After your Instant Answer goes live for the World to use, you will be able to see [traffic information](http://docs.duckduckhack.com/maintaining/tools.html#monitor-your-instant-answer-page) about your IA as well as receive feedback directly from users. 
                
Please refer to the [maintenance guide and docs](http://docs.duckduckhack.com/maintaining/guidelines.html) to keep up with potential changes to your Instant Answer.
                
Thank you again for your help!

**Comment by bfmags** [comment]
Great! Thanks :)

File: lib/fathead/react_js/parse.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        """
+        Extract the Example Code Snippet
+        """
+        example_code = ''
+        for tag in api.next_siblings:
+            if tag.name == 'div':
+                code = str(tag.find('pre'))
+                if code != 'None':
+                    example_code = code
+        return example_code.strip('\n')
```
Comment (dip-kush) [inline lib/fathead/react_js/parse.py]: Sorry for missing this. I will fix it
@MariagraziaAlastra  Are there any more mistakes ?

**Comment by tagawa** [comment]
Hi @dip-kush. Just wanted to let you know this is now live on DuckDuckGo!
https://duckduckgo.com/?q=React.Component&ia=about&iax=1

You can also see a chart showing coverage and engagement in the forum topic:
https://forum.duckduckhack.com/t/reactjs-fathead/999

Thanks again for creating this for JS developers 😃

**Pull request merged**